l3kernel/l3sort.dtx

```diff
@@ -968,31 +1021,641 @@
+    \cs_new:Npn \@@_quick_single_end:nnnwnw #1#2#3#4 \s_@@_mark #5#6 \s_@@_stop
+      { #5 {#3} #6 \s_@@_stop }
+    \cs_new:Npn \@@_quick_only_i_end:nnnwnw #1#2#3#4 \s_@@_mark #5#6 \s_@@_stop
+      {
+        \@@_quick_split:NnNn #1
+          \@@_quick_end:nnTFNn { } \s_@@_mark {#5}
+        {#3}
+        #6 \s_@@_stop
+      }
+    \cs_new:Npn \@@_quick_only_ii_end:nnnwnw #1#2#3#4 \s_@@_mark #5#6 \s_@@_stop
+      {
+        \@@_quick_split:NnNn #2
+          \@@_quick_end:nnTFNn { } \s_@@_mark { #5 {#3} }
+        #6 \s_@@_stop
+      }
+    \cs_new:Npn \@@_quick_split_end:nnnwnw #1#2#3#4 \s_@@_mark #5#6 \s_@@_stop
+      {
+        \@@_quick_split:NnNn #2 \@@_quick_end:nnTFNn { } \s_@@_mark
+          {
+            \@@_quick_split:NnNn #1
+              \@@_quick_end:nnTFNn { } \s_@@_mark {#5}
+            {#3}
+          }
+        #6 \s_@@_stop
+      }
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+%    \begin{macrocode}
+\cs_if_exist:NT \tex_expanded:D
   {
-    \@@_quick_split:NnNn #1
-      \@@_quick_end:nnTFNn { } \s_@@_mark {#5}
-    {#3}
-    #6 \s_@@_stop
+%    \end{macrocode}
+%
+% \begin{macro}[EXP]
+%   {\@@_if_stop:w, \@@_if_mark:w, \@@_if_tail:w}
+%   Since the three scan marks are forbidden to appear in the lists we can setup
+%   quick tests by just gobbling everything up until that marker.
+%    \begin{macrocode}
+    \cs_new:Npn \@@_if_stop:w #1 \s_@@_merge_exp_stop {}
+    \cs_new:Npn \@@_if_mark:w #1 \s_@@_merge_exp_mark {}
+    \cs_new:Npn \@@_if_tail:w #1 \s_@@_merge_exp_tail {}
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}[rEXP]
+%   {
+%     \@@_merge_exp_bottom_up_tl:Nnn,
+%     \@@_merge_exp_bottom_up_tl_mark:w
+%     \@@_merge_exp_bottom_up_tl_stop:w
+%   }
+%   The first bottom up pass grabs two items and sorts them. Markers are placed
+%   to indicate the end of each sublist. Since we grab two items in each
+%   recursion step we have to check both of these values for the end of list
+%   markers. If the input token list has an odd number of elements, |#3| will be
+%   \cs{s_@@_merge_exp_mark} in the last recursion. In this case |#2| will be
+%   the last element and form a sublist with just one element. If the input
+%   token list has an even number of elements, |#3| will be
+%   \cs{s_@@_merge_exp_stop} in the last recursion and we just leave the markers
+%   for the end of sublists.
+%
+%   After the first bottom up pass every sublist will be sorted and followed by
+%   \cs{s_@@_merge_exp_tail}. The sublists are separated by
+%   \cs{s_@@_merge_exp_list}, and the end of sublists is indicated by a sublist
+%   containing only \cs{s_@@_merge_exp_mark} and another one containing
+%   \cs{s_@@_merge_exp_stop}.
+%    \begin{macrocode}
+    \cs_new:Npn \@@_merge_exp_bottom_up_tl:Nnn #1#2#3
+      {
+        \@@_if_stop:w #3 \@@_merge_exp_bottom_up_tl_stop:w \s_@@_merge_exp_stop
+        \@@_if_mark:w #3 \@@_merge_exp_bottom_up_tl_mark:w \s_@@_merge_exp_mark
```

Comment:
The usual trick to speed up the loop is to have two end-markers (like `{a}{b}{c}\s_mark \s_mark \s_stop`) and then test whether the `#3` argument is the end-marker.  Then you can refine to distinguish even/odd by testing the `#2` argument.  Actually, if you have two end-markers like this, you should be able to do both tests in one go more cleanly as `\@@_if_mark:w #2 \@@_merge_even:w #3 \@@_merge_odd:w \s_@@_merge_exp_mark`.

I have to say, your use of `\@@_if_mark:w` and similar to end the loop is really nice in comparison to our much slower quark tests in `\tl_map_function:nN`.